src/libs/API/commands/BaseAPICommandBlocking.js

```diff
@@ -0,0 +1,44 @@
+import BaseAPICommand from './BaseAPICommand';
+import Onyx from 'react-native-onyx';
+import ONYXKEYS from '../../../ONYXKEYS';
+import CONST from '../../../CONST';
+
+/**
+ * API Commands that need to have some kind of blocking UI while the network request is being made
+ * can extend this class.
+ */
+export default class extends BaseAPICommand {
+    // Use withOnyx to access the data about this blocking request (eg. ValidateEmailCommand.blockingCommandOnyxKey)
+    static blockingCommandOnyxKey = `BlockingCommand-${this.commandName}`;
```

Comment:
> `Command_ValidateEmail.key`

I like this, but what is the `.key` for?

> Is there a difference between an "action" and "command"?

I totally get where you're going with this. In the first iteration of this I actually just called `(new ValidateEmailCommand()).makeRequest(accountID, validateCode);` directly from the view and removed `User.ValidateEmail()`;

In the end, all the "action" would be doing is calling a "command", and all the logic is inside the "command". In that sense then, the "action" becomes completely unnecessary.

Reply:
>I like this, but what is the .key for?

Maybe `onxyKey` at least to set the intent that we can subscribe to it and get information about the state of request(s)

Reply:
>In the end, all the "action" would be doing is calling a "command", and all the logic is inside the "command". In that sense then, the "action" becomes completely unnecessary.

Exactly. But we do have actions that "just set Onyx data" and don't call any APIs.

Reply:
I think where this breaks down with actions and commands is that an "action" can be a group of commands. I am primarily thinking of the sign-in flow where we do many things like creating temporary logins, deleting temporary logins, etc.

However, the more I think about that, the more I wonder if that should all just be moved to PHP.

The original intention of actions was to manage Onyx data. We end up with actions that:
1. Only manage Onyx data and make no network requests
2. Make a single network request and store the data in Onyx
3. Make multiple network requests and store the data in Onyx

I think that 2 and 3 are basically the same, so maybe it's more like we have `Action` and `ActionWithNetworkRequests`.

I don't know. Maybe this is splitting hairs and we should just stick with "Action" for everything.

Reply:
I think I agree with moving that all to PHP (though it would require new and more specific API commands). Synchronous chains of API requests could get complex using the new system and it would be easier to handle single blocking commands instead of several.

If we do that, then "actions manage Onyx data" and "commands trigger reads and writes" sounds like clear guidance.

Reply:
Yeah, the only problem though is that these commands will also be managing onyx data, so that's not a clear separation.

What's really happening is that `API.js` is going away, and we're essentially getting an individual file for each thing that used to be a single method in `API.js`, but also has a lot of logic that used to be in the action. So, I think of it like:

1. Views always call actions
5. Commands make network requests and manage the responses
3. Actions can manage Onyx data directly (if no network requests are necessary) or 
4. Actions can call trigger a command

I would also be open to renaming "command" to something different. `APICommand` `NetworkCommand` `APIRequest` `API`

Reply:
What about something like this...

1. Views always call actions
2. Actions come in two flavors... "client actions" and "network actions"
3. Network actions make network requests and manage the responses
4. Client actions modify data without making any network requests
5. Both network actions and client actions can manage Onyx data directly

We could switch that up to be "local actions" or something else but I would be fine with leaning into "it's all actions"

Reply:
so yeah I guess 

>Maybe this is splitting hairs and we should just stick with "Action" for everything.

👍

Reply:
oooooooo I like where this is going. The "client actions" and "network actions" gave me an idea. We could have:

1. A class for `Action` that allows any action to "manage data in onyx"
2. A subclass for `NetworkAction` that also allows for making the network requests

Then there would be several subclasses of `NetworkAction` like `NetworkActionBlocking` and `NetworkActionOptimistic`. (or then maybe we don't really need `NetworkAction` at all)